In [ ]:
import pandas as pd

df = pd.read_csv('/content/BostonHousing.csv')
display(df.head())

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

Check for missing values in each column and then impute the missing values in the 'rm' column with the mean, as indicated by the previous `df.info()` output. Finally, verify that the missing values have been handled.



In [ ]:
print("Missing values before imputation:")
print(df.isnull().sum())
# Impute missing values in 'rm' with the mean
df['rm'] = df['rm'].fillna(df['rm'].mean())
print("\nMissing values after imputation:")
print(df.isnull().sum())

Missing values before imputation:
crim       0
zn         0
indus      0
chas       0
nox        0
rm         5
age        0
dis        0
rad        0
tax        0
ptratio    0
b          0
lstat      0
medv       0
dtype: int64

Missing values after imputation:
crim       0
zn         0
indus      0
chas       0
nox        0
rm         0
age        0
dis        0
rad        0
tax        0
ptratio    0
b          0
lstat      0
medv       0
dtype: int64


In [ ]:
import numpy as np

numerical_cols = df.select_dtypes(include=np.number).columns
for col in numerical_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    df[col] = np.where(df[col] < lower_bound, lower_bound, df[col])
    df[col] = np.where(df[col] > upper_bound, upper_bound, df[col])
display(df.describe())

crim          zn       indus   chas         nox          rm  \
count  506.000000  506.000000  506.000000  506.0  506.000000  501.000000   
mean     2.225032    6.963439   11.136779    0.0    0.554695    6.275094   
std      3.313353   12.028788    6.860353    0.0    0.115878    0.633620   
min      0.006320    0.000000    0.460000    0.0    0.385000    4.772500   
25%      0.082045    0.000000    5.190000    0.0    0.449000    5.884000   
50%      0.256510    0.000000    9.690000    0.0    0.538000    6.208000   
75%      3.677083   12.500000   18.100000    0.0    0.624000    6.625000   
max      9.069639   31.250000   27.740000    0.0    0.871000    7.736500   

              age         dis         rad         tax     ptratio           b  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.783947    9.549407  408.237154   18.463834  381.918836   
std     28.148861    2.069765    8.707259  168.537116    2.143924   19.054913   
min      2.900000    1.129600    1.000000  187.000000   13.200000  344.106250   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000    9.820800   24.000000  711.000000   22.000000  396.900000   

            lstat        medv  
count  506.000000  506.000000  
mean    12.612011   21.877075  
std      7.016829    7.602976  
min      1.730000    5.062500  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     31.962500   36.962500

In [ ]:
from sklearn.preprocessing import StandardScaler
import numpy as np

numerical_cols = df.select_dtypes(include=np.number).columns
print("Numerical columns identified:")
print(numerical_cols)

Numerical columns identified:
Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')


In [ ]:
scaler = StandardScaler()
df[numerical_cols] = scaler.fit_transform(df[numerical_cols])
display(df.describe())

crim            zn         indus   chas           nox  \
count  5.060000e+02  5.060000e+02  5.060000e+02  506.0  5.060000e+02   
mean   1.123388e-16  1.228705e-16  2.106352e-16    0.0 -1.965929e-16   
std    1.000990e+00  1.000990e+00  1.000990e+00    0.0  1.000990e+00   
min   -6.702900e-01 -5.794707e-01 -1.557842e+00    0.0 -1.465882e+00   
25%   -6.474129e-01 -5.794707e-01 -8.676906e-01    0.0 -9.130288e-01   
50%   -5.947057e-01 -5.794707e-01 -2.110985e-01    0.0 -1.442174e-01   
75%    4.386757e-01  4.607314e-01  1.015999e+00    0.0  5.986790e-01   
max    2.067808e+00  2.021035e+00  2.422565e+00    0.0  2.732346e+00   

                 rm           age           dis           rad         tax  \
count  5.060000e+02  5.060000e+02  5.060000e+02  5.060000e+02  506.000000   
mean  -1.000517e-15 -1.474446e-16  1.965929e-16 -1.123388e-16    0.000000   
std    1.000990e+00  1.000990e+00  1.000990e+00  1.000990e+00    1.000990   
min   -2.372654e+00 -2.335437e+00 -1.283708e+00 -9.828429e-01   -1.313990   
25%   -6.200418e-01 -8.374480e-01 -8.143137e-01 -6.379618e-01   -0.767576   
50%   -1.029626e-01  3.173816e-01 -2.788081e-01 -5.230014e-01   -0.464673   
75%    5.483661e-01  9.067981e-01  6.792404e-01  1.661245e+00    1.530926   
max    2.300978e+00  1.117494e+00  2.919572e+00  1.661245e+00    1.798194   

            ptratio             b         lstat          medv  
count  5.060000e+02  5.060000e+02  5.060000e+02  5.060000e+02  
mean  -6.459479e-16 -1.165515e-15 -2.808469e-17  2.808469e-16  
std    1.000990e+00  1.000990e+00  1.000990e+00  1.000990e+00  
min   -2.457663e+00 -1.986365e+00 -1.552379e+00 -2.213767e+00  
25%   -4.966998e-01 -3.436284e-01 -8.077173e-01 -6.388126e-01  
50%    2.736785e-01  5.001642e-01 -1.786063e-01 -8.914209e-02  
75%    8.106088e-01  7.515291e-01  6.195515e-01  4.111568e-01  
max    1.651022e+00  7.869881e-01  2.760455e+00  1.986111e+00

*   The initial dataset contained 506 entries and 14 columns.
*   The 'rm' column was identified to have 5 missing values, which were successfully imputed using the mean of the column.
*   Outliers in numerical columns were detected using the IQR method and handled by capping values at the calculated lower and upper bounds.
*   The dataset did not contain any categorical features, so no categorical encoding was performed.
*   All numerical features were successfully scaled using standardization, resulting in features with a mean close to 0 and a standard deviation close to 1.